src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+    (** Maps an index int the {!val:vars} array to a slot. *) }
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action. *)
+  type t =
+    { clauses : rule list
+    (** The rules. *)
+    ; var_met : int
+    (** Number of variables encountered so far. *) }
+
+  (** [pp o m] prints matrix [m] to out channel [o]. *)
+  let pp : t pp = fun oc { clauses ; _ } ->
+    let module F = Format in
+    let pp_line oc l =
+      F.fprintf oc "@[<h>" ;
+      F.pp_print_list ~pp_sep:(fun _ () -> F.fprintf oc ";@ ")
+        pp_component oc (Array.to_list l) ;
+      F.fprintf oc "@]" in
+    F.fprintf oc "{@[<v>@," ;
+    F.pp_print_list ~pp_sep:F.pp_print_cut pp_line oc
+      (List.map (fun { lhs ; _ } -> lhs) clauses) ;
+    F.fprintf oc "@.}@,"
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : Terms.rule list -> t = fun rs ->
+    let r2r r =
+      let term_pos = List.to_seq r.Terms.lhs |> Subterm.tag |> Array.of_seq in
+      { lhs = term_pos ; rhs = r.Terms.rhs ; env_builder = [] } in
+    { clauses = List.map r2r rs ; var_met = 0 }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> m.clauses = []
+
+  (** [get_col n m] retrieves column [n] of matrix [m]. *)
+  let get_col : int -> t -> component list = fun ind { clauses ; _ } ->
+    List.map (fun { lhs ; _ } -> lhs.(ind)) clauses
+
+  (** [score c] returns the score heuristic for column [c]. *)
+  let rec score : component list -> int = function
+    | []                              -> 0
+    | (x, _) :: xs when is_treecons x -> score xs
+    | _ :: xs                         -> succ (score xs)
+
+  (** [pick_best_among m c] returns the index of the best column of matrix [m]
+      among columns [c] according to a heuristic. *)
+  let pick_best_among : t -> int array -> int = fun mat columns->
+    let wild_pc = Array.map (fun ci ->
+      let c = get_col ci mat in score c) columns in
+    Array.argmax (<=) wild_pc
+
+  (** [yield m] yields a rule to be applied. *)
+  let yield : t -> rule option = fun { clauses ; _ } ->
+    let is_exhausted { lhs ; _ } =
+      Array.for_all (fun (e, _) -> not (is_treecons e)) lhs in
+    List.find_opt is_exhausted clauses
+
+  (** [can_switch_on p k] returns whether a switch can be carried out on
+      column [k] in matrix [p] *)
+  let can_switch_on : t -> int -> bool = fun { clauses ; _ } k ->
+    let rec loop : rule list -> bool = function
+      | []      -> false
+      | r :: rs -> if is_treecons (fst r.lhs.(k)) then true else loop rs in
+    loop clauses
+
+  (** [discard_cons_free m] returns the list of indexes of columns containing
+      terms that can be matched against (discard constructor-free columns). *)
+  let discard_cons_free : t -> int array = fun m ->
+    let ncols = List.extremum (>)
+      (List.map (fun { lhs ; _ } -> Array.length lhs) m.clauses) in
```

Comment:
List.map could be removed by replacing (>) by its composition with the argument of List.map.